estimagic/decorators.py

```diff
@@ -47,23 +47,78 @@ def wrapper_numpy_interface(x, *args, **kwargs):
     return decorator_numpy_interface
 
 
+def negative_criterion(criterion):
+    """Turn maximization into minimization."""
+
+    @functools.wraps(criterion)
+    def wrapper_negative_criterion(*args, **kwargs):
+        out = criterion(*args, **kwargs)
+        if np.isscalar(out):
+            criterion_value = -out
+            comparison_plot_data = pd.DataFrame({"value": [np.nan]})
+        else:
+            criterion_value, comparison_plot_data = -out[0], out[1]
+
+        return criterion_value, comparison_plot_data
+
+    return wrapper_negative_criterion
+
+
 def log_evaluation(database, tables):
-    """Log parameters and fitness values."""
+    """Log parameters and fitness values.
+
+    The criterion function is allowed to have two returns:
 
-    def decorator_log_parameters_and_criterion_value(func):
+    1. The criterion value which can be a scalar or an array, e.g., moments in MSM.
+    2. Additional data for the comparison plot in the tidy data format.
+
+    """
+
+    def decorator_log_evaluation(func):
         @functools.wraps(func)
-        def wrapper_log_parameters_and_criterion_value(params, *args, **kwargs):
-            criterion_value = func(params, *args, **kwargs)
+        def wrapper_log_evaluation(params, *args, **kwargs):
+            out = func(params, *args, **kwargs)
+
+            if np.isscalar(out):
+                criterion_value = out
+                cp_data = {"value": pd.DataFrame({"value": [np.nan]})}
+            else:
+                criterion_value, comparison_plot_data = out
+                cp_data = {"value": comparison_plot_data["value"].to_numpy()}
```

Comment:
You handle several times the "expansion" from one to two outputs (e.g. in also in lines 56-61). Why not have one `expand_output` decorator and "concatenate"? This might also allow us to avoid case handling elsewhere (e.g. in `optimize` lines 310-320?).

Reply:
Done.